File: src/domain/reminders/reminder.ts
import type { CalendarAccount } from "../calendar-account.js";
import type { NamedFailure, Outcome } from "../failure.js";
import { failed, succeeded } from "../failure.js";
import type { LocalDay } from "../time-zone.js";

/**
 * A container of reminders in one calendar account. Titles repeat across accounts; the
 * identifier doesn't.
 */
export type ReminderList = {
  readonly identifier: string;
  readonly title: string;
  readonly account: CalendarAccount;
};

/**
 * When a reminder is due: a due date, which is a whole day with no time of day, or a due time,
 * which is an instant. A due date is never given a time, not even midnight (upstream #34).
 */
export type Due = { readonly day: LocalDay } | { readonly at: Date };

/** An item in a reminder list. It is open until it is completed. */
export type Reminder = {
  readonly identifier: string;
  readonly title: string;
  readonly isCompleted: boolean;
  readonly due?: Due;
  readonly reminderList: ReminderList;
};

/**
 * Reminders in the order a reader expects them: open ones before completed ones, so nothing
 * done reads as still to do, then by title. Reminders with one title fall back to their
 * identifier, so the same reminders always read the same way and a page never repeats or skips
 * one.
 */
export const openBeforeCompleted = (reminders: readonly Reminder[]): readonly Reminder[] =>
  [...reminders].sort(
    (left, right) =>
      Number(left.isCompleted) - Number(right.isCompleted) ||
      left.title.localeCompare(right.title) ||
      left.identifier.localeCompare(right.identifier),
  );

/** How many reminders a page holds when the caller names no limit, and the most it may name. */
export const defaultLimit = 100;
export const greatestLimit = 500;

/** One page of reminders, and the offset of the next when there is one. */
export type Page = {
  readonly reminders: readonly Reminder[];
  readonly nextOffset?: number;
};

/** The reminders from an offset, up to a limit. They must already be in reading order. */
export const pageOf = (reminders: readonly Reminder[], offset: number, limit: number): Page => {
  const end = offset + limit;
  return {
    reminders: reminders.slice(offset, end),
    ...(reminders.length > end ? { nextOffset: end } : {}),
  };
};

/** What a reminder list's name comes to among the reminder lists there are. */
export type ReminderListNamed =
  | { readonly kind: "one"; readonly list: ReminderList }
  | { readonly kind: "none" }
  | { readonly kind: "several"; readonly lists: readonly ReminderList[] };

/**
 * The reminder list a caller means, by its identifier or else by its exact title. A title two
 * accounts share names several, and choosing between them is the caller's to do, never ours.
 */
export const reminderListNamed = (
  reminderLists: readonly ReminderList[],
  name: string,
): ReminderListNamed => {
  const byIdentifier = reminderLists.find(({ identifier }) => identifier === name);
  if (byIdentifier) return { kind: "one", list: byIdentifier };

  const byTitle = reminderLists.filter(({ title }) => title === name);
  const [only] = byTitle;
  if (byTitle.length === 1 && only) return { kind: "one", list: only };

  return byTitle.length === 0 ? { kind: "none" } : { kind: "several", lists: byTitle };
};

/**
 * What was asked for and the lists that answer to it, as a record: a title is text the user
 * wrote, so it is a field and never part of a sentence (ADR-0006).
 */
const evidence = (name: string, reminderLists: readonly ReminderList[]): string =>
  JSON.stringify({ asked: name, reminderLists });

/**
 * The one reminder list a name means, or the refusal saying why there isn't one, which begins
 * with what did not happen: nothing is ever read from, or created in, a list it guessed at.
 */
export const theReminderListNamed = (
  reminderLists: readonly ReminderList[],
  name: string,
  whatDidNotHappen: string,
): Outcome<ReminderList> => {
  const named = reminderListNamed(reminderLists, name);
  const refusal = (
    code: string,
    sentence: string,
    lists: readonly ReminderList[],
  ): NamedFailure => ({
    code,
    sentence: `${whatDidNotHappen} ${sentence}`,
    evidence: evidence(name, lists),
  });

  switch (named.kind) {
    case "one":
      return succeeded(named.list);
    case "none":
      return failed(
        refusal(
          "reminder-list-unknown",
          "No reminder list has that identifier or title.",
          reminderLists,
        ),
      );
    case "several":
      return failed(
        refusal(
          "reminder-list-ambiguous",
          "Reminder lists in more than one account have that title, so ask by identifier.",
          named.lists,
        ),
      );
  }
};

File: native/Sources/HelperCore/Reminder.swift
import Foundation

/// A container of reminders in one account, addressed by its identifier because titles repeat
/// across accounts.
public struct ReminderList: Equatable, Hashable, Sendable {
  public let identifier: String
  public let title: String
  public let account: CalendarAccount

  public init(identifier: String, title: String, account: CalendarAccount) {
    self.identifier = identifier
    self.title = title
    self.account = account
  }
}

/// What a read of some reminder lists found. The ones that did not answer in time are named
/// rather than silently missing, so a short answer is never mistaken for the whole; the ones
/// deleted since they were listed are named apart, because they are gone rather than slow.
public struct RemindersRead: Equatable, Sendable {
  public let reminders: [Reminder]
  public let unreadReminderLists: [ReminderList]
  public let goneReminderLists: [ReminderList]

  public init(
    reminders: [Reminder], unreadReminderLists: [ReminderList],
    goneReminderLists: [ReminderList] = []
  ) {
    self.reminders = reminders
    self.unreadReminderLists = unreadReminderLists
    self.goneReminderLists = goneReminderLists
  }
}

/// When a reminder is due: a due date is a day with no time of day, and a due time is an
/// instant. A due date is never given a time, not even midnight (upstream #34, #64).
public enum Due: Equatable, Sendable {
  case date(year: Int, month: Int, day: Int)
  case time(Date)
}

/// An item in a reminder list. It is open until it is completed.
public struct Reminder: Equatable, Sendable {
  public let identifier: String
  public let title: String
  /// Searched, and never sent: an index keeps to small fixed fields (ADR-0006).
  public let notes: String?
  public let isCompleted: Bool
  public let due: Due?
  public let reminderList: ReminderList

  public init(
    identifier: String, title: String, notes: String?, isCompleted: Bool, due: Due?,
    reminderList: ReminderList
  ) {
    self.identifier = identifier
    self.title = title
    self.notes = notes
    self.isCompleted = isCompleted
    self.due = due
    self.reminderList = reminderList
  }
}

extension Reminder {
  /// Whether the title or the notes mention some text, ignoring case: both are upper-cased, then
  /// compared code unit by code unit, which the server's fake store does the same way (Greek's
  /// final sigma folds alike upper-cased, and lower-casing folds it by position). The text is
  /// compared and never interpreted, so it matches only itself.
  func mentions(_ text: String) -> Bool {
    let sought = text.uppercased()
    return [title, notes].contains {
      $0?.uppercased().range(of: sought, options: .literal) != nil
    }
  }
}

/// A reminder that does not exist yet, and the reminder list it is to be created in. It carries no
/// alert, and there is no way to give it one: a due time says when it is due (#20).
public struct NewReminder: Equatable, Sendable {
  public let title: String
  public let reminderListIdentifier: String
  public let due: Due?
  /// Saved as given, and never sent back: an index keeps to small fixed fields (ADR-0006).
  public let notes: String?

  public init(title: String, reminderListIdentifier: String, due: Due?, notes: String? = nil) {
    self.title = title
    self.reminderListIdentifier = reminderListIdentifier
    self.due = due
    self.notes = notes
  }
}

/// A reminder as the store holds it, and how many alerts it holds on it.
public struct HeldReminder: Equatable, Sendable {
  public let reminder: Reminder
  public let alerts: Int

  public init(reminder: Reminder, alerts: Int) {
    self.reminder = reminder
    self.alerts = alerts
  }
}

/// What became of a new reminder. Confirmed means the store showed it afterwards; unconfirmed
/// means it was saved and could not be found again, which is never a failure to retry.
struct CreatedReminder: Equatable, Sendable {
  let reminder: Reminder
  let confirmed: Bool
  /// How many alerts the store holds on it, known only when it could be read back.
  let alerts: Int?
  /// Whether the store holds notes on it, known only when it could be read back.
  let hasNotes: Bool?
}
